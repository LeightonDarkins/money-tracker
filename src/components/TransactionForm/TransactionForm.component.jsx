import React from 'react'
import PropTypes from 'prop-types'
import './TransactionForm.scss'

const TransactionForm = ({
  amount,
  onAmountChange,
  date,
  onDateChange,
  category,
  onCategoryChange,
  categories,
  account,
  onAccountChange,
  accounts,
  onSubmitClick,
  onCancelClick,
  transactionType,
  onTypeChange
}) => {
  const categoryOptions = () => {
    return categories.map(category => (
      <option key={category.id} value={category.id}>{ category.name }</option>
    ))
  }

  const accountOptions = () => {
    return accounts.map(account => (
      <option key={account.id} value={account.id}>{ account.name }</option>
    ))
  }

  const onInputFocus = (e) => {
    e.target.select()
  }

  const localAmountChange = (e) => {
    const numberToSend = Number((e.target.value * 100).toFixed(0))

    return isNaN(numberToSend) ? null : onAmountChange(numberToSend)
  }

  return (
    <div className='transaction-form'>
      <h1>Create Transaction</h1>
      <div>
        <div>
          <label>Type</label>
          <select value={transactionType} onChange={onTypeChange}>
            <option value='expense'>Expense</option>
            <option value='income'>Income</option>
          </select>
        </div>
        <label>Amount</label>
        <div className='mt-number-input-container'>
          <input type='number'
            value={amount / 100}
            onChange={localAmountChange}
            onFocus={onInputFocus} />
        </div>
      </div>

      <div>
        <label>Date</label>
        <input value={date} onChange={onDateChange} />
      </div>

      <div>
        <label>Category</label>
        <select value={category} onChange={onCategoryChange}>
          { categoryOptions() }
        </select>
      </div>

      <div>
        <label>Account</label>
        <select value={account} onChange={onAccountChange}>
          { accountOptions() }
        </select>
      </div>

      <div className='transaction-form-actions'>
        <button className='cancel' onClick={() => onCancelClick()}>
          Cancel
        </button>
        <button className='confirm' onClick={() => onSubmitClick({ amount, date, category, account, transactionType })}>
          Done
        </button>
      </div>
    </div>
  )
}

TransactionForm.propTypes = {
  amount: PropTypes.number.isRequired,
  onAmountChange: PropTypes.func.isRequired,
  date: PropTypes.string.isRequired,
  onDateChange: PropTypes.func.isRequired,
  category: PropTypes.string.isRequired,
  onCategoryChange: PropTypes.func.isRequired,
  categories: PropTypes.array.isRequired,
  account: PropTypes.string.isRequired,
  onAccountChange: PropTypes.func.isRequired,
  accounts: PropTypes.array.isRequired,
  onSubmitClick: PropTypes.func.isRequired,
  onCancelClick: PropTypes.func.isRequired,
  transactionType: PropTypes.string.isRequired,
  onTypeChange: PropTypes.func.isRequired
}

export default TransactionForm
