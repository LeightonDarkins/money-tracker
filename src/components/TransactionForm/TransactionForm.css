label {
  font-family: Roboto;
  font-size: 13px;
  display: block;
  margin: 5px 0 5px 0;
}

input {
  height: 41px;
  width: 310px;
  padding: 0 5px 0 5px;
  margin: 5px 0 5px 0;
  font-size: 20px;
}

select {
  height: 41px;
  width: 310px;
  margin: 5px 0 5px 0;
}

button {
  height: 41px;
  width: 310px;
  margin: 5px 0 5px 0;
}
