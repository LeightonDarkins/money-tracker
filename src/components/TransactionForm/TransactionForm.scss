@import '../../variables.scss';
@import '~react-datepicker/dist/react-datepicker.css';

.transaction-form {
  width: 100%;

  label {
    font-size: 13px;
    display: block;
    margin: 5px 0 5px 0;
  }

  input {
    padding: 0 5px 0 5px;
  }

  input, select {
    width: 310px;
    font-size: $default-font-size;
    margin: 5px;
    padding-left: 5px;
  }

  button, select, input {
    height: $default-element-height;
  }

  .mt-number-input-container {
    position: relative;

    input {
      text-align: right;
    }
  }

  .mt-number-input-container:before {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    font-size: $default-font-size;
    content: '$';
    left: 10px;
  }
}

.transaction-form-actions {
  display: flex;
  justify-content: center;
  width: 100%;

  button {
    width: 100%;
    margin: 5px;
  }

  .confirm {
    color: #fff;
    background-color: #08B477;
    font-size: 16px;
    border: none;
  }

  .confirm:hover {
    color: #eee;
    cursor: pointer;
    background-color: #51EBB4;
  }

  .cancel {
    color: #fff;
    background-color: #B42501;
    font-size: 16px;
    border: none;
  }

  .cancel:hover {
    color: #eee;
    cursor: pointer;
    background-color: #EB6848;
  }
}
