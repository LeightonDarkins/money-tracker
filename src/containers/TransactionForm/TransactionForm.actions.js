export const types = {
  amountChanged: 'TRANSACTION_AMOUNT_CHANGED',
  dateChanged: 'TRANSACTION_DATE_CHANGED',
  accountChanged: 'TRANSACTION_ACCOUNT_CHANGED',
  categoryChanged: 'TRANSACTION_CATEGORY_CHANGED',
  formSubmitted: 'TRANSACTION_FORM_SUBMITTED',
  formCancelled: 'TRANSACTION_FORM_CANCELLED',
  typeChanged: 'TRANSACTION_TYPE_CHANGED',
  transactionFormFetchAccountsSucceeded: 'TRANSACTION_FORM_FETCH_ACCOUNTS_SUCCEEDED',
  transactionFormFetchCategoriesSucceeded: 'TRANSACTION_FORM_FETCH_CATEGORIES_SUCCEEDED',
  clearTransactionForm: 'CLEAR_TRANSACTION_FORM',
  returnDefaultDate: 'RETURN_DEFAULT_DATE'
}

export const amountChanged = (amount) => {
  return {
    type: types.amountChanged,
    amount
  }
}

export const dateChanged = (date) => {
  return {
    type: types.dateChanged,
    date
  }
}

export const accountChanged = (account) => {
  return {
    type: types.accountChanged,
    account
  }
}

export const categoryChanged = (category) => {
  return {
    type: types.categoryChanged,
    category
  }
}

export const formSubmitted = (transactionDetails) => {
  return {
    type: types.formSubmitted,
    transactionDetails
  }
}

export const formCancelled = () => {
  return { type: types.formCancelled }
}

export const typeChanged = (transactionType) => {
  return {
    type: types.typeChanged,
    transactionType
  }
}

export const returnDefaultDate = (date) => {
  return {
    type: types.returnDefaultDate,
    date
  }
}

export const transactionFormFetchAccountsSucceeded = (accounts) => {
  return { type: types.transactionFormFetchAccountsSucceeded, accounts }
}

export const transactionFormFetchCategoriesSucceeded = (categories) => {
  return { type: types.transactionFormFetchCategoriesSucceeded, categories }
}

export const clearTransactionForm = () => {
  return { type: types.clearTransactionForm }
}
