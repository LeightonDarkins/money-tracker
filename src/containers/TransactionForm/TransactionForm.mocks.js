import moment from 'moment'

module.exports = {
  accounts: [
    {
      id: 'account-1',
      name: 'Account One'
    },
    {
      id: 'account-2',
      name: 'Account Two'
    }
  ],

  categories: [
    {
      id: 'category-1',
      name: 'Category One'
    },
    {
      id: 'category-2',
      name: 'Category Two'
    }
  ],

  formattedToday: moment().format('DD/MM/YYYY'),

  date: moment('31-01-2001', 'DD-MM-YYYY')
}
