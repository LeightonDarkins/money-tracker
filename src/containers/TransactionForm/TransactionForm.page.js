module.exports = {
  getAmountInputElement: (wrapper) => {
    return wrapper.find('.mt-number-input-container input')
  },

  getDatePickerElement: (wrapper) => {
    return wrapper.find('.mt-datepicker')
  },

  getAccountSelectorElement: (wrapper) => {
    return wrapper.find('.mt-account-select')
  },

  getAccountOptions: (selector) => {
    return selector.find('.mt-account-option')
  },

  getCategorySelectorElement: (wrapper) => {
    return wrapper.find('.mt-category-select')
  },

  getTypeSelectorElement: (wrapper) => {
    return wrapper.find('.mt-type-select')
  }
}
