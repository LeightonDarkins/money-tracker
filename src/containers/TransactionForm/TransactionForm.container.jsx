import { connect } from 'react-redux'
import {
  amountChanged,
  dateChanged,
  accountChanged,
  categoryChanged,
  formSubmitted,
  formCancelled
} from './TransactionForm.actions'
import TransactionForm from '../../components/TransactionForm/TransactionForm.component.jsx'

const mapStateToProps = (state, ownProps) => {
  return {
    amount: state.TransactionForm.amount,
    date: state.TransactionForm.date,
    category: state.TransactionForm.category,
    categories: state.TransactionForm.categories,
    account: state.TransactionForm.account,
    accounts: state.TransactionForm.accounts
  }
}

const mapDispatchToProps = dispatch => {
  return {
    onSubmitClick: (accountDetails) => {
      dispatch(formSubmitted(accountDetails))
    },
    onCancelClick: () => {
      dispatch(formCancelled())
    },
    onAmountChange: (event) => {
      dispatch(amountChanged(event.target.value))
    },
    onDateChange: (event) => {
      dispatch(dateChanged(event.target.value))
    },
    onAccountChange: (event) => {
      dispatch(accountChanged(event.target.value))
    },
    onCategoryChange: (event) => {
      dispatch(categoryChanged(event.target.value))
    }
  }
}

const ConnectedTransactionForm = connect(mapStateToProps, mapDispatchToProps)(TransactionForm)

export default ConnectedTransactionForm
