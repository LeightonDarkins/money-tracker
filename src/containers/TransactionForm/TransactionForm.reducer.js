import { types } from './TransactionForm.actions'
import moment from 'moment'

const initialState = { amount: 0, date: moment(), category: '', account: '', accounts: [], categories: [], transactionType: 'expense' }

const hasContents = (array) => array.length > 0

const categoryForm = (state = initialState, action) => {
  switch (action.type) {
    case types.amountChanged:
      return Object.assign({}, state, {
        amount: action.amount
      })

    case types.dateChanged:
      return Object.assign({}, state, {
        date: action.date
      })

    case types.categoryChanged:
      return Object.assign({}, state, {
        category: action.category
      })

    case types.accountChanged:
      return Object.assign({}, state, {
        account: action.account
      })

    case types.typeChanged:
      return Object.assign({}, state, {
        transactionType: action.transactionType
      })

    case types.transactionFormFetchAccountsSucceeded:
      if (hasContents(action.accounts)) {
        return Object.assign({}, state, {
          account: action.accounts[0].id,
          accounts: action.accounts
        })
      }

      return Object.assign({}, state, {
        accounts: action.accounts
      })

    case types.transactionFormFetchCategoriesSucceeded:
      if (hasContents(action.categories)) {
        return Object.assign({}, state, {
          category: action.categories[0].id,
          categories: action.categories
        })
      }

      return Object.assign({}, state, {
        categories: action.categories
      })

    case types.clearTransactionForm:
      return Object.assign({}, state, initialState)

    case types.returnDefaultDate:
      return Object.assign({}, state, {
        date: action.date
      })

    default:
      return state
  }
}

export default categoryForm
