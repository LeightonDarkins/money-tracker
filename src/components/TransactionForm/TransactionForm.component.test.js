/* global describe, it, expect, jest, beforeEach, afterEach */

import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'
import TransactionForm from './TransactionForm.component'
import moment from 'moment'
import th from '../../common/TestHelpers'
import DatePicker from 'react-datepicker'

const testDate = moment('19/10/2017', 'DD/MM/YYYY')

const defaultProps = {
  amount: 0,
  onAmountChange: th.empty,
  date: testDate,
  onDateChange: th.empty,
  category: 'category-1',
  onCategoryChange: th.empty,
  categories: [{ id: 'category-1' }, { id: 'category-2' }],
  account: 'account-1',
  onAccountChange: th.empty,
  accounts: [{ id: 'account-1' }, { id: 'account-2' }],
  onSubmitClick: th.empty,
  onCancelClick: th.empty,
  transactionType: 'expense',
  onTypeChange: th.empty,
  startDate: testDate,
  fetchCategories: th.empty,
  fetchAccounts: th.empty
}

const mockProps = (props) => {
  let mockedProps = props

  mockedProps.onAmountChange = jest.fn()
  mockedProps.onDateChange = jest.fn()
  mockedProps.onCategoryChange = jest.fn()
  mockedProps.onAccountChange = jest.fn()
  mockedProps.onSubmitClick = jest.fn()
  mockedProps.onCancelClick = jest.fn()
  mockedProps.onTypeChange = jest.fn()
  mockedProps.fetchCategories = jest.fn()
  mockedProps.fetchAccounts = jest.fn()

  return mockedProps
}

const resetMocks = (props) => {
  props.onAmountChange.mockReset()
  props.onDateChange.mockReset()
  props.onCategoryChange.mockReset()
  props.onAccountChange.mockReset()
  props.onSubmitClick.mockReset()
  props.onCancelClick.mockReset()
  props.onTypeChange.mockReset()
  props.fetchCategories.mockReset()
  props.fetchAccounts.mockReset()
}

describe('TransactionForm', () => {
  describe('Snapshots', () => {
    it('should render correctly in default state', () => {
      const transactionForm = renderer.create(<TransactionForm {...defaultProps} />)

      expect(transactionForm).toMatchSnapshot()
    })

    it('should render placeholders when data is missing', () => {
      const transactionForm = renderer.create(<TransactionForm
        {...defaultProps}
        categories={[]}
        accounts={[]} />)

      expect(transactionForm).toMatchSnapshot()
    })
  })

  describe('Mount', () => {
    it('calls fetchAccounts and fetchCategories when rendered', () => {
      const mockedProps = mockProps(defaultProps)
      mount(<TransactionForm {...mockedProps} />)

      expect(mockedProps.fetchAccounts).toBeCalled()
      expect(mockedProps.fetchCategories).toBeCalled()
    })
  })

  describe('Behaviour', () => {
    let wrapper, mockedProps

    beforeEach(() => {
      mockedProps = mockProps(defaultProps)
      wrapper = shallow(<TransactionForm {...mockedProps} />)
    })

    afterEach(() => {
      resetMocks(mockedProps)
    })

    it('calls onAmountChange, with a value in cents, when the amount is changed', () => {
      const changeEvent = {
        target: {
          value: '10'
        }
      }

      wrapper.find('.mt-number-input-container input').simulate('change', changeEvent)

      expect(mockedProps.onAmountChange).toBeCalledWith(1000)
    })

    it('does not call onAmountChange when the amount is not a number', () => {
      const changeEvent = {
        target: {
          value: 'test'
        }
      }

      wrapper.find('.mt-number-input-container input').simulate('change', changeEvent)

      expect(mockedProps.onAmountChange).not.toBeCalled()
    })

    it('calls onDateChange when the date is changed', () => {
      const input = '10/19/2017'

      wrapper.find(DatePicker).simulate('change', input)

      expect(mockedProps.onDateChange).toBeCalledWith(input)
    })

    it('calls onCategoryChange when the category is changed', () => {
      const selector = wrapper.find('.mt-category-select')

      selector.simulate('change', 'category-2')

      expect(mockedProps.onCategoryChange).toBeCalledWith('category-2')
    })

    it('calls onAccountChange when the account is changed', () => {
      const selector = wrapper.find('.mt-account-select')

      selector.simulate('change', 'account-2')

      expect(mockedProps.onAccountChange).toBeCalledWith('account-2')
    })

    it('calls onTypeChange when the type is changed', () => {
      const selector = wrapper.find('.mt-type-select')

      selector.simulate('change', 'income')

      expect(mockedProps.onTypeChange).toBeCalledWith('income')
    })

    it('calls onSubmitClick when the submit button is clicked', () => {
      wrapper.find('.confirm').simulate('click')

      const expectedFields = {
        account: mockedProps.account,
        amount: mockedProps.amount,
        category: mockedProps.category,
        date: mockedProps.date,
        transactionType: mockedProps.transactionType
      }

      expect(mockedProps.onSubmitClick).toBeCalledWith(expectedFields)
    })

    it('calls onCancelClick when the cancel button is clicked', () => {
      wrapper.find('.cancel').simulate('click')

      expect(mockedProps.onCancelClick).toBeCalled()
    })
  })
})
