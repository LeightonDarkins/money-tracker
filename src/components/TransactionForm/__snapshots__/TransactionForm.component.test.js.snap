exports[`TransactionForm Snapshots should render correctly in default state 1`] = `
<div
  id="TransactionForm">
  <section
    className="mt-header">
    <h2>
      Create Transaction
    </h2>
  </section>
  <section
    className="mt-body">
    <div>
      <div>
        <label>
          Type
        </label>
        <select
          className="mt-type-select"
          onChange={[Function]}
          value="expense">
          <option
            value="expense">
            Expense
          </option>
          <option
            value="income">
            Income
          </option>
        </select>
      </div>
      <label>
        Amount
      </label>
      <div
        className="mt-number-input-container">
        <input
          onChange={[Function]}
          onFocus={[Function]}
          type="number"
          value={0} />
      </div>
    </div>
    <div>
      <label>
        Date
      </label>
      <div>
        <div
          className="react-datepicker-wrapper">
          <div
            className="react-datepicker__input-container">
            <input
              autoComplete={undefined}
              autoFocus={undefined}
              className="mt-datepicker"
              disabled={false}
              id={undefined}
              name={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder={undefined}
              readOnly={undefined}
              required={undefined}
              tabIndex={undefined}
              title={undefined}
              type="text"
              value="19/10/2017" />
          </div>
        </div>
      </div>
    </div>
    <div>
      <label>
        Category
      </label>
      <select
        className="mt-category-select"
        onChange={[Function]}
        value="category-1">
        <option
          className="mt-category-option"
          value="category-1" />
        <option
          className="mt-category-option"
          value="category-2" />
      </select>
    </div>
    <div>
      <label>
        Account
      </label>
      <select
        className="mt-account-select"
        onChange={[Function]}
        value="account-1">
        <option
          className="mt-account-option"
          value="account-1" />
        <option
          className="mt-account-option"
          value="account-2" />
      </select>
    </div>
  </section>
  <section
    className="mt-footer transaction-form-actions">
    <button
      className="cancel"
      onClick={[Function]}>
      Cancel
    </button>
    <button
      className="confirm"
      onClick={[Function]}>
      Done
    </button>
  </section>
</div>
`;

exports[`TransactionForm Snapshots should render placeholders when data is missing 1`] = `
<div
  id="TransactionForm">
  <section
    className="mt-header">
    <h2>
      Create Transaction
    </h2>
  </section>
  <section
    className="mt-body">
    <div>
      <div>
        <label>
          Type
        </label>
        <select
          className="mt-type-select"
          onChange={[Function]}
          value="expense">
          <option
            value="expense">
            Expense
          </option>
          <option
            value="income">
            Income
          </option>
        </select>
      </div>
      <label>
        Amount
      </label>
      <div
        className="mt-number-input-container">
        <input
          onChange={[Function]}
          onFocus={[Function]}
          type="number"
          value={0} />
      </div>
    </div>
    <div>
      <label>
        Date
      </label>
      <div>
        <div
          className="react-datepicker-wrapper">
          <div
            className="react-datepicker__input-container">
            <input
              autoComplete={undefined}
              autoFocus={undefined}
              className="mt-datepicker"
              disabled={false}
              id={undefined}
              name={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder={undefined}
              readOnly={undefined}
              required={undefined}
              tabIndex={undefined}
              title={undefined}
              type="text"
              value="19/10/2017" />
          </div>
        </div>
      </div>
    </div>
    <div>
      <label>
        Category
      </label>
      <select
        className="mt-category-select"
        onChange={[Function]}
        value="category-1">
        <option
          className="mt-category-option"
          disabled={true}>
          No Categories Available
        </option>
      </select>
    </div>
    <div>
      <label>
        Account
      </label>
      <select
        className="mt-account-select"
        onChange={[Function]}
        value="account-1">
        <option
          className="mt-account-option"
          disabled={true}>
          No Accounts Available
        </option>
      </select>
    </div>
  </section>
  <section
    className="mt-footer transaction-form-actions">
    <button
      className="cancel"
      onClick={[Function]}>
      Cancel
    </button>
    <button
      className="confirm"
      onClick={[Function]}>
      Done
    </button>
  </section>
</div>
`;
